File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node==null) return node;
        HashMap<Node, Node> visited= new HashMap<>();
        return dfs(node, visited);
    }

    private Node dfs(Node node, HashMap<Node, Node> visited){
        if(visited.containsKey(node)){
            return visited.get(node);
        }
        Node copy= new Node(node.val);
        visited.put(node,copy);
        for(Node neighbor: node.neighbors){
            copy.neighbors.add(dfs(neighbor,visited));
        }
        return copy;
    }
}

File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        visited={}
        if node is None:
            return node
        return self.dfs(node, visited)
    def dfs(self, node, visited):
        if node in visited:
            return visited[node]
        copy=  Node(node.val)
        visited[node]=copy

        for neighbor in node.neighbors:
            copy.neighbors.append(self.dfs(neighbor,visited))
        return copy        